src/OpenApi/src/OpenApiGenerator.cs

```diff
@@ -0,0 +1,474 @@
+        {
+            return GetOperation(method, methodInfo, metadata, pattern);
+        }
+
+        return null;
+    }
+
+    private OpenApiOperation GetOperation(string httpMethod, MethodInfo methodInfo, EndpointMetadataCollection metadata, RoutePattern pattern)
+    {
+        var disableInferredBody = ShouldDisableInferredBody(httpMethod);
+        return new OpenApiOperation
+        {
+            OperationId = metadata.GetMetadata<IEndpointNameMetadata>()?.EndpointName,
+            Summary = metadata.GetMetadata<IEndpointSummaryMetadata>()?.Summary,
+            Description = metadata.GetMetadata<IEndpointDescriptionMetadata>()?.Description,
+            Tags = GetOperationTags(methodInfo, metadata),
+            Parameters = GetOpenApiParameters(methodInfo, metadata, pattern, disableInferredBody),
+            RequestBody = GetOpenApiRequestBody(methodInfo, metadata, pattern),
+            Responses = GetOpenApiResponses(methodInfo, metadata)
+        };
+
+        static bool ShouldDisableInferredBody(string method)
+        {
+            // GET, DELETE, HEAD, CONNECT, TRACE, and OPTIONS normally do not contain bodies
+            return method.Equals(HttpMethods.Get, StringComparison.Ordinal) ||
+                   method.Equals(HttpMethods.Delete, StringComparison.Ordinal) ||
+                   method.Equals(HttpMethods.Head, StringComparison.Ordinal) ||
+                   method.Equals(HttpMethods.Options, StringComparison.Ordinal) ||
+                   method.Equals(HttpMethods.Trace, StringComparison.Ordinal) ||
+                   method.Equals(HttpMethods.Connect, StringComparison.Ordinal);
+        }
+    }
+
+    private static OpenApiResponses GetOpenApiResponses(MethodInfo method, EndpointMetadataCollection metadata)
+    {
+        var responses = new OpenApiResponses();
+        var responseType = method.ReturnType;
+        if (AwaitableInfo.IsTypeAwaitable(responseType, out var awaitableInfo))
+        {
+            responseType = awaitableInfo.ResultType;
+        }
+
+        if (typeof(IResult).IsAssignableFrom(responseType))
+        {
+            responseType = typeof(void);
+        }
+
+        var errorMetadata = metadata.GetMetadata<ProducesErrorResponseTypeAttribute>();
+        var defaultErrorType = errorMetadata?.Type;
+
+        var responseProviderMetadata = metadata.GetOrderedMetadata<IApiResponseMetadataProvider>();
+        var producesResponseMetadata = metadata.GetOrderedMetadata<IProducesResponseTypeMetadata>();
+
+        var eligibileAnnotations = new Dictionary<int, (Type?, MediaTypeCollection)>();
+
+        foreach (var responseMetadata in producesResponseMetadata)
+        {
+            var statusCode = responseMetadata.StatusCode;
+
+            var discoveredTypeAnnotation = responseMetadata.Type;
+            var discoveredContentTypeAnnotation = new MediaTypeCollection();
+
+            if (discoveredTypeAnnotation == typeof(void))
```

Comment:
Is this needed? Looks like we check null or void shortly below and set to responseType then.

Reply:
Yeah, this isn't needed. It exists as a legacy from the ApiExplorer days where `== typeof(void)` was used to handle scenarios where the return type was not provided by the user via an attribute, in which case, we would only fall back to the defined responseType on a 2xx status code.

However, that was coupled with some weirdness in the way that content types were handled in 

Anyhoo, I was going to clean this up in a follow-up commit had I gotten any feedback in addition to this on the PR but I'll do this as part of https://github.com/dotnet/aspnetcore/issues/41243 instead.